src/access/wac.ts

```diff
@@ -68,6 +84,25 @@ function aclAccessToUniversal(access: AclAccess): WacAccess {
   } as WacAccess;
 }
 
+function universalAccessToAcl(access: WacAccess): AclAccess {
+  // Universal access is aligned on ACP, which means there is a distinction between
+  // controlRead and controlWrite. This split doesn't exist in WAC, which is why
+  // the type for the input variable of this function is a restriction on the
+  // universal Access type. Also, in WAC, an undefined and false Access modes are
+  // equivalent.
+  if (access.controlRead !== access.controlWrite) {
+    throw new Error(
+      "For WAC resources, controlRead and controlWrite must be equal."
+    );
+  }
+  return {
+    read: access.read ? true : false,
+    append: access.append ? true : false,
+    write: access.write ? true : false,
+    control: access.controlRead ? true : false,
```

Comment:
Actually, I just realised something: when you pass `false` for a property to `setAgentResourceAccessWac`, it does remove existing statements that give the user access. So I think what we should be doing here is allow for both `undefined` and a `boolean`, and have `undefined` result in whatever the existing value is. "Denying" access has somewhat different semantics than it has with ACPs, but it's still distinct from leaving things unchanged.

(And actually, that makes me realise that the ACP API might be a bit troublesome either: although you can e.g. give someone read access, you can't _remove_ that read access anymore - you can only _deny_ read access. This might be more than you want. So I'm wondering if maybe we should align the ACP API with the WAC one, and have a value of `false` only remove existing `acp:allow` values, rather than adding `acp:deny` ones?

Reply:
So that means the API could have a `setAuthorizedAgentAccess` and `setDeniedAgentAccess`, each accepting modes with `true` to set it, `false` to remove it, and `undefined` to leave it as it is. An alternative would be to change `Access` to become something like
```
type Access = {
  read: "add" | "remove" | "deny" | undefined
  write: ...
}
```
which would result in not splitting the access-related functions between authorize and deny in the API, but looks less convenient to use.

Reply:
Yeah that would be cumbersome. The "problem" with adding separate allow and deny functions, is that the latter doesn't work for WAC at all, so it's not really universal. So maybe we should just change `setAgentAccess`'s behaviour to what you called `setAuthorizedAgentAccess`? (Which is kinda confusing given that we also have `setAuthorizedAccess` :) )

Reply:
Oh dear I didn't think about `setAuthorizedAccess`.

And actually even for ACP only having `setAgentAccess` is limiting: if I `setAgentAccess(resource, agent, {read:false})`, do I intend to deny access to the agent, or to remove read access granted individually to the agent if it exists, but they could still have read access through a Group (which isn't the case if the access is denied I think) ? While if we have `setAgentAuthorizedAccess` and `setAgentDeniedAccess` (name WIP), the latter approach is applied more explicitly ? And when talking to an ACL backend, `setAgentDeniedAccess` would simply apply a reverse logic compared to `setAgentAuthorizedAccess`: remove access modes marked `true`, add triples marked `false`; and leave the undefined ones.

Reply:
> `setAgentDeniedAccess` would simply apply a reverse logic compared to `setAgentAuthorizedAccess`: remove access modes marked true, add triples marked false; and leave the undefined ones.

What do you mean by "add triples marked false"? That sounds like specifying deny, which isn't possible in WAC?

I'd think you could simply pass e.g. `read: true` to `setAgentAccess` which _adds_ an allow statement, `read: false` to remove any existing ones, and `undefined` to leave existing ones be. That is also the current behaviour of the WAC-specific API, IIRC.

Reply:
> >  if I setAgentAccess(resource, agent, {read:false}), do I intend to deny access to the agent, or to remove read access granted individually to the agent if it exists, but they could still have read access through a Group (which isn't the case if the access is denied I think) ?

> It would remove read access granted explicitly to the given agent, and they could indeed still have read access through a group. We explicitly make specify the caveat that we cannot guarantee effective access, because there is no way for us to do that :)

If I follow, `setAgentAccess` (from the universal API) would add specified access on `true`, remove specified access on `false`, and leave the given access untouched on `undefined`. That means that effectively, we don't support denying access through the universal API. 

> > setAgentDeniedAccess would simply apply a reverse logic compared to setAgentAuthorizedAccess: remove access modes marked true, add triples marked false; and leave the undefined ones.

> What do you mean by "add triples marked false"? That sounds like specifying deny, which isn't possible in WAC?
I'd think you could simply pass e.g. read: true to setAgentAccess which adds an allow statement, read: false to remove any existing ones, and undefined to leave existing ones be. That is also the current behaviour of the WAC-specific API, IIRC.

What I meant is in WAC, basically removing a granted access and denying access are two equivalent things. Therefore, if I `setAgentDeniedAccess(resource, agent, {read: true})`, in the case of an ACL backend that would simply remove any triples giving Read access to the Resource for the Agent, being strictly equivalent to `setAgentAuthorizedAccess(resource, agent, {read: false})`. In the case of an ACP backend though, these two statements would not have a similar effect on the effective access the Agent gets to the Resource.

I'm happy to have `setAgentAccess` of the first proposal behave the same as `setAgentAuthorizedAccess` of the second proposal, but then I'm not sure how `deny` access could ever be set.

Reply:
Right, so indeed my proposal would be to not allow setting `deny` access using the universal API, because that functionality is not common to all access systems we support - just like we don't support setting Creator access using this API.

So we basically have two options:
- Only support allow (`true`), unallow (`false`) and leave the same (`undefined`) - i.e. no deny.
- Have allow- and deny-specific functions, where the former is as above, and the latter supports deny (`true`), undeny (`false`) and leave the same (`undefined`) in the case of ACP, and unallow (`true`), allow (`false`) and leave the same (`undefined`) in the case of WAC.

As mentioned I'm inclined to the former, but not wedded to it and see the point of the latter. Should be bring this up in the team?

Reply:
Whichever approach we chose, 7afa637110544b89b9f12c33f4cff175a7f863e2 fixes the current behaviour and accepts `boolean | undefined` for each mode.

Reply:
For the record, after some discussion, we'll be going with the 'no deny' approach to keep the universal API as universal as possible, and limit the addition of features that are access control system-specific.